x/slashing/keeper.go

```diff
@@ -123,7 +132,13 @@ func (k Keeper) handleValidatorSignature(ctx sdk.Context, addr crypto.Address, p
 			// Downtime confirmed: slash and jail the validator
 			logger.Info(fmt.Sprintf("Validator %s past min height of %d and below signed blocks threshold of %d",
 				pubkey.Address(), minHeight, k.MinSignedPerWindow(ctx)))
-			k.validatorSet.Slash(ctx, consAddr, height, power, k.SlashFractionDowntime(ctx))
+			// We need to retrieve the stake distribution which signed the block, so we subtract ValidatorUpdateDelay from the evidence height,
+			// and subtract an additional 1 since this is the LastCommit.
+			// Note that this *can* result in a "distributionHeight" of -1 or -2,
+			// i.e. at the end of the pre-genesis block (none) = at the beginning of the genesis block.
+			// That's fine since this is just used to filter unbonding delegations & redelegations.
```

Comment:
So most of these changes look fine (superficially), however I find it a bit confusing to review - I'd imagine that if we were somehow able to abstract delay logic it would make the code more maintainable. - Doesn't have to block merge, but just something to keep thinking about

Reply:
I don't think there's a way to abstract delay logic more on *this* line, since the `LastCommit` being an additional block offset is just how ABCI works - but I think it can be done elsewhere.